protocol/backend/eth_util.go

```diff
@@ -0,0 +1,54 @@
+package backend
+
+import (
+	"context"
+	"time"
+
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/ethclient"
+)
+
+// SleepWithContext is the same as time.Sleep(...) but with preemption if the context is complete.
+func SleepWithContext(ctx context.Context, d time.Duration) {
+	timer := time.NewTimer(d)
+	select {
+	case <-ctx.Done():
+		if !timer.Stop() {
+			<-timer.C
+		}
+	case <-timer.C:
+	}
+}
+
+// WaitForEthBlockAfterTimestamp returns the header of the first block received where the block's timestamp
+// is >= ts.
+func WaitForEthBlockAfterTimestamp(ctx context.Context, ec *ethclient.Client, ts int64) (*ethtypes.Header, error) {
+	timeDelta := time.Duration(ts-time.Now().Unix()) * time.Second
```

Comment:
```suggestion
func WaitForEthBlockAfterTimestamp(ctx context.Context, ec *ethclient.Client, ts time.Time) (*ethtypes.Header, error) {
	timeDelta := time.Until(ts).Seconds()
```

I think this should be equivalent?